### PR TITLE
feat(dead_code,dead_symbols): count_test_imports, barrel tracing & root entry points

### LINKED ISSUE #35: Dead code detector should flag code used only in tests
## Problem

Currently, the dead code detector does NOT flag functions and variables as dead code when they are only referenced in test files. However, code that exists only to support testing and is never used in production code could be considered dead code (or at least should be configurable to be treated as such).

## Current Behavior

- A function/variable defined in source code
- Only used/imported in test files (e.g., `*.test.ts`, `*.spec.ts`, `__tests__/*`)
- NOT used anywhere in production code
- Does NOT get flagged as dead code

## Expected Behavior

Code that is used ONLY in tests and never in production should be flagged as dead code, or there should be a configuration option to control this behavior.

## Rationale

1. **Test-only code might indicate design issues** - If a function is only used in tests, it might mean the abstraction is unnecessary
2. **Maintenance burden** - Test-only functions still need to be maintained even though they provide no production value
3. **Code bloat** - Increases bundle size for no production benefit
4. **Configurability** - Different teams have different policies about test-only code

## Possible Solutions

1. **Add configuration option** - `flag_test_only_code: bool` to control whether test-only usage should be considered dead
2. **Separate category** - Flag as `test_only_code` instead of `dead_code` for clarity
3. **Exclude test helpers** - Allow patterns to exclude legitimate test utilities (e.g., `src/test-utils/*`)

## Example

```typescript
// src/utils.ts
export function testHelper() { 
  // Only used in tests, never in production
}

// src/utils.test.ts
import { testHelper } from './utils';

test('something', () => {
  testHelper();
});
```

Currently `testHelper` is NOT flagged, but arguably it should be (with proper configuration).

## Related

This relates to the `detect_test_files` functionality and could use similar test file pattern detection.

### LINKED ISSUE #37: Dead symbol detector should trace re-exports through barrel files
## Problem

When a symbol is re-exported through a barrel file (`index.ts`), the dead_code detector considers the originating file as "alive" because it is re-exported by an entry point. However, the re-exported symbol itself may never be imported by any consumer.

This means symbols that are only used in tests but re-exported through barrel files are invisible to both:
- `dead_code` detector (file is re-exported → not dead)
- `count_test_imports: false` option (file-level check bypassed by re-export)

## Example

```
packages/utils/src/
  index.ts          ← entry point, barrel: export * from './payload-sanitizer.js'
  payload-sanitizer.ts  ← exports sanitizePayload()
  payload-sanitizer.spec.ts ← only consumer of sanitizePayload()
```

`sanitizePayload` is never imported by any production code, only by the test file. But the dead_code detector sees that `payload-sanitizer.ts` is re-exported through `index.ts` (an entry point) and considers it alive.

## Current Behavior

- `dead_code` detector works at **file level**
- `is_reexported()` returns `true` because `index.ts` re-exports `payload-sanitizer.ts`
- File is not flagged, even though no consumer ever imports `sanitizePayload` from the barrel

## Expected Behavior

The dead symbol detector (`dead_symbols`) should trace re-exports through barrel files and verify that the re-exported symbol is actually imported by at least one consumer downstream. If no consumer imports the symbol (only tests do), it should be flagged.

## Possible Approaches

1. **Enhance `dead_symbols` detector** — when a symbol is re-exported through a barrel, check if any file imports that symbol from the barrel (not just that the barrel re-exports it)
2. **Transitive usage analysis** — follow re-export chains and verify end-to-end usage
3. **Combine with `count_test_imports`** — if the only downstream consumers are test files, flag the symbol

## Related

- #35 — `count_test_imports` option (file-level, doesn't cover barrel re-exports)

### PR DESCRIPTION
**Pull request by fixcik**
## Summary

- **`count_test_imports` option** for `dead_code` and `dead_symbols` detectors — code used only in tests is flagged as dead by default (`false`); set to `true` to restore old behavior
- 17 built-in test file patterns (`.test.ts`, `.spec.ts`, `__tests__/`, etc.)
- **Barrel re-export tracing** for `dead_symbols`: `export * from` is no longer treated as actual usage — traces through barrel chains to verify the symbol reaches a real consumer
- **Root entry point logic**: only entry points belonging to the scanned project's own `package.json` whitelist symbols as public API; sub-package entry points in monorepos are traced normally
- **Tiered verbose levels**: `-v` for debug (summaries), `-vv` for trace (per-symbol tracing)
- Fix `Config.ignore` serde default: test file patterns are now preserved when loading from YAML config files

## Configuration

```yaml
rules:
  dead_code:
    count_test_imports: false  # default
  dead_symbols:
    count_test_imports: false  # default
    exclude:
      - packages/testing/**  # exclude test utility packages
```

## Test plan

- [x] Integration test: `count_test_imports=false` flags test-only code
- [x] Integration test: `count_test_imports=true` keeps test-only code
- [x] Integration test: default behavior flags test-only code
- [x] Test file itself is not flagged (entry point)
- [x] Production code not affected
- [x] Barrel re-export tracing: symbols re-exported via `export *` are not falsely marked alive
- [x] Barrel tracing handles cycles (visited set)
- [x] Root entry points: monorepo scan detects dead exports in sub-packages
- [x] Root entry points: standalone package scan treats public API as alive
- [x] `-v` shows debug logs, `-vv` shows trace logs, no flags = clean output
- [x] Config ignore defaults preserved when YAML config is loaded
- [x] Clippy clean, schema updated

Closes #35
Closes #37

---

> [!NOTE]
> **Medium Risk**
> Changes core detector logic for dead code/symbol detection (including monorepo entry-point handling), which can shift results and affect CI gating; new tests and explicit option defaults reduce risk but regressions in edge cases are still possible.
> 
> **Overview**
> Improves `dead_code`/`dead_symbols` accuracy by adding a `count_test_imports` option (default *false*) so code referenced only from test files can be flagged as dead, while allowing opt-in to count test imports as usage.
> 
> `dead_symbols` now distinguishes *consumption* from barrel re-exports (`export * from`) by tracing through barrel chains (cycle-safe) and only treating re-exported symbols as public API when they flow through **root** entry points (derived from discovered `package.json` directories for monorepos). This also introduces tiered CLI verbosity (`-v` debug, `-vv` trace) and centralizes default test ignore patterns in config/schema, with new fixtures/tests covering the new behaviors.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 13cdad29015823ff4c93c3732455afd1b8122515. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: crates/archlint/src/detectors/hygiene/dead_code.rs [hunk 1]
```diff
@@ -66,6 +66,13 @@ impl Detector for DeadCodeDetector {
         }
         combined_exclude.extend_from_slice(&ctx.config.ignore);
 
+        // When false, imports from test files are not counted as usage,
+        // so code used only in tests is flagged as dead code.
+        let count_test_imports: bool = rule.get_option("count_test_imports").unwrap_or(false);
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_code.rs]: we broke backward-compatible it's ok

File: crates/archlint/src/detectors/hygiene/dead_code.rs [hunk 2]
```diff
@@ -66,6 +66,13 @@ impl Detector for DeadCodeDetector {
...
+        if !count_test_imports {
+            combined_exclude.extend(Self::TEST_FILE_PATTERNS.iter().map(ToString::to_string));
+        }
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_code.rs]: Breaking backward compatibility is intentional. Default changed to false to avoid flagging production code used only in tests as 'not dead' by default.

File: crates/archlint/src/detectors/hygiene/dead_symbols.rs [hunk 3]
```diff
@@ -25,20 +25,72 @@ struct MethodCheckContext<'a> {
...
+    "**/*.e2e-spec.ts",
+    "**/*.e2e-spec.js",
+    "**/__tests__/**",
+    "**/__mocks__/**",
+    "**/test/**",
+    "**/tests/**",
+    "**/__fixtures__/**",
+    "**/*.mock.ts",
+    "**/*.mock.js",
+];
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_symbols.rs]: Breaking backward compatibility is intentional - count_test_imports defaults to false by design.

File: crates/archlint/src/config/types.rs [hunk 4]
```diff
@@ -438,22 +441,26 @@ pub struct LayerConfig {
...
+        "**/*.spec.js".to_string(),
+        "**/__tests__/**".to_string(),
+        "**/__mocks__/**".to_string(),
+        "**/test/**".to_string(),
+        "**/tests/**".to_string(),
+        "**/__fixtures__/**".to_string(),
+        "**/*.mock.ts".to_string(),
+        "**/*.mock.js".to_string(),
+    ]
+}
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/config/types.rs:465]: Fixed: Added missing .tsx/.jsx and .e2e-spec.* patterns to TEST_FILE_PATTERNS. Centralized all test patterns in types.rs to eliminate duplication.

[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/config/types.rs:465]: Fixed: Added missing tsx/jsx/e2e test patterns. Created centralized TEST_FILE_PATTERNS constant in types.rs (17 patterns).

File: crates/archlint/src/config/types.rs [hunk 5]
```diff
@@ -7,7 +7,10 @@ use std::collections::HashMap;
 /// Defines project settings, rules, and framework extensions.
 #[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
 pub struct Config {
-    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    #[serde(
+        default = "default_ignore_patterns",
+        skip_serializing_if = "Vec::is_empty"
+    )]
     pub ignore: Vec<String>,
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/config/types.rs]: Fixed: Removed skip_serializing_if attribute so explicit ignore: [] is preserved. Users can now opt out of default patterns by setting an empty list.

File: crates/archlint/src/detectors/hygiene/dead_code.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_code.rs:90]: Fixed: count_test_imports=true now works correctly. When true, test file patterns are filtered from config.ignore before adding to combined_exclude, allowing test imports to be counted as usage.

File: crates/archlint/src/detectors/hygiene/dead_symbols.rs [hunk 6]
```diff
@@ -181,9 +267,36 @@ impl DeadSymbolsDetector {
...
+                        return true;
+                    }
+                } else {
+                    // Actual namespace import (import * as X): check if symbol name is used
+                    if let Some(syms) = file_symbols.get(importer) {
+                        if syms.local_usages.contains(symbol_name) {
+                            return true;
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_symbols.rs:371]: Fixed: check_dead_local_symbols now skips test files when count_test_imports=false to avoid false positives for local symbols used within the same test file.

File: crates/archlint/src/detectors/hygiene/dead_code.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_code.rs]: Breaking backward compatibility is intentional - see author's response above.

File: crates/archlint/src/detectors/hygiene/dead_symbols.rs [hunk 7]
```diff
@@ -1,13 +1,23 @@
...
 /// Initializes the detector module.
 /// This function is used for module registration side-effects.
 pub const fn init() {}
 
+static COMPILED_TEST_PATTERNS: LazyLock<Vec<glob::Pattern>> = LazyLock::new(|| {
+    TEST_FILE_PATTERNS
+        .iter()
+        .filter_map(|p| glob::Pattern::new(p).ok())
+        .collect()
+});
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_symbols.rs]: Fixed in commit 7558964: replaced LazyLock with once_cell::sync::Lazy for MSRV 1.75 compatibility.

File: crates/archlint/src/detectors/hygiene/dead_symbols.rs [hunk 8]
```diff
@@ -56,6 +124,7 @@ impl DeadSymbolsDetector {
         all_smells.extend(Self::check_dead_exports(
             file_symbols,
             entry_points,
+            &root_entry_points,
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_symbols.rs:137]: Fixed in commit 8991f26: check_dead_exports and check_dead_methods now skip test files via test_excluded set when count_test_imports=false.

File: crates/archlint/src/detectors/hygiene/dead_symbols.rs [hunk 9]
```diff
@@ -56,6 +131,7 @@ impl DeadSymbolsDetector {
...
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_symbols.rs:137]: Duplicate of the issue above — fixed in commit 8991f26: test files are now skipped in check_dead_exports and check_dead_methods.

File: crates/archlint/src/detectors/hygiene/dead_symbols.rs [hunk 10]
```diff
@@ -168,22 +252,116 @@ impl DeadSymbolsDetector {
...
+        let mut visited = HashSet::new();
+        Self::is_symbol_consumed_recursive(
+            symbol_usages,
+            file_symbols,
+            entry_points,
+            file_path,
+            symbol_name,
+            &mut visited,
+        )
+    }
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_symbols.rs:279]: Fixed in commit 6b0372a: renamed entry_points → root_entry_points across is_symbol_consumed, is_symbol_consumed_recursive, and check_wildcard_importers.

File: crates/archlint/src/detectors/hygiene/dead_symbols.rs [hunk 11]
```diff
@@ -168,22 +256,116 @@ impl DeadSymbolsDetector {
...
+                    symbol_name,
+                    visited,
+                ) {
+                    return true;
+                }
+            } else if let Some(syms) = file_symbols.get(importer) {
+                // Actual namespace import (import * as X): check if symbol name is used
+                if syms.local_usages.contains(symbol_name) {
+                    return true;
+                }
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/src/detectors/hygiene/dead_symbols.rs:371]: Fixed in commit 3987eed: removed the else-if exclusion so local_usages check runs after barrel tracing regardless of is_reexport. Now a file that both re-exports and namespace-imports from the same source correctly detects consumption.

File: crates/archlint/tests/dead_symbols_barrel_test.rs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        }),
+    );
+
+    Config {
+        rules,
+        entry_points: vec!["main.ts".to_string()],
+        ignore: Vec::new(),
+        ..Default::default()
+    }
+}
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/tests/dead_symbols_barrel_test.rs:56]: Fixed in commit 13cdad2: added the same inline comment explaining why ignore: Vec::new() is used.

File: crates/archlint/tests/dead_symbols_barrel_test.rs [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+    let detector = archlint::detectors::dead_symbols::DeadSymbolsDetector;
+    let smells = detector.detect(&ctx);
+
+    assert!(
+        is_dead_symbol(&smells, "testHelper"),
+        "testHelper should be flagged: re-exported through barrel but never consumed. Got smells: {:?}",
+        smells.iter().filter_map(|s| if let archlint::detectors::SmellType::DeadSymbol { name, .. } = &s.smell_type { Some(name.as_str()) } else { None }).collect::<Vec<_>>()
+    );
+}
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/tests/dead_symbols_barrel_test.rs:84]: Fixed in commit 3987eed: the test helper now resolves import sources to absolute paths (matching production EngineBuilder::resolve_symbols), so build_symbol_imports_map keys correctly match is_symbol_consumed_recursive lookups. Barrel tracing is now properly exercised in tests.

File: crates/archlint/tests/dead_symbols_barrel_test.rs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    let ctx = analyze_fixture_with_config("dead_symbols_barrel", config);
+
+    let detector = archlint::detectors::dead_symbols::DeadSymbolsDetector;
+    let smells = detector.detect(&ctx);
+
+    assert!(
+        !is_dead_symbol(&smells, "testHelper"),
+        "testHelper should NOT be flagged when count_test_imports=true"
+    );
+}
```
[the remark this answers is not in the document]
Comment (fixcik) [reply crates/archlint/tests/dead_symbols_barrel_test.rs:116]: Fixed in commit 13cdad2: added test_barrel_default_flags_test_only_usage test that exercises the default behavior (no explicit count_test_imports) end-to-end.
